x-pack/plugins/lens/public/datatable_visualization/sorting.tsx

```diff
@@ -0,0 +1,91 @@
+  if (!ipaddr.isValid(ip)) {
+    // for non valid IPs have the same behaviour as for now (we assume it's only the "Other" string)
+    // create a mock object which has all a special value to keep them always at the bottom of the list
+    return { parts: Array(8).fill(directionFactor * Infinity) };
+  }
+  const parsedIp = ipaddr.parse(ip);
+  return isIPv6Address(parsedIp) ? parsedIp : parsedIp.toIPv4MappedAddress();
+}
+
+function getIPCriteria(sortBy: string, directionFactor: number) {
+  // Create a set of 8 function to sort based on the 8 IPv6 slots of an address
+  // For IPv4 bring them to the IPv6 "mapped" format and then sort
+  return (rowA: Record<string, unknown>, rowB: Record<string, unknown>) => {
+    const ipAString = rowA[sortBy] as string;
+    const ipBString = rowB[sortBy] as string;
+    const ipA = getSafeIpAddress(ipAString, directionFactor);
+    const ipB = getSafeIpAddress(ipBString, directionFactor);
+
+    // Now compare each part of the IPv6 address and exit when a value != 0 is found
+    let i = 0;
+    let diff = ipA.parts[i] - ipB.parts[i];
+    while (!diff && i < 7) {
+      i++;
+      diff = ipA.parts[i] - ipB.parts[i];
+    }
+
+    // in case of same address but written in different styles, sort by string length
+    if (diff === 0) {
+      return directionFactor * (ipAString.length - ipBString.length);
+    }
+    return directionFactor * diff;
+  };
+}
+
+function getRangeCriteria(sortBy: string, directionFactor: number) {
+  // fill missing fields with these open bounds to perform number sorting
+  const openRange = { gte: -Infinity, lt: Infinity };
+  return (rowA: Record<string, unknown>, rowB: Record<string, unknown>) => {
+    const rangeA = { ...openRange, ...(rowA[sortBy] as Omit<Range, 'type'>) };
+    const rangeB = { ...openRange, ...(rowB[sortBy] as Omit<Range, 'type'>) };
+
+    const fromComparison = rangeA.gte - rangeB.gte;
+    const toComparison = rangeA.lt - rangeB.lt;
+
+    return directionFactor * (fromComparison || toComparison);
+  };
+}
+
+export function getSortingCriteria(
+  type: string | undefined,
+  sortBy: string,
+  formatter: FieldFormat,
+  direction: string
+) {
+  // handle the direction with a multiply factor.
+  const directionFactor = direction === 'asc' ? 1 : -1;
+
+  if (['number', 'date'].includes(type || '')) {
+    return (rowA: Record<string, unknown>, rowB: Record<string, unknown>) =>
+      directionFactor * ((rowA[sortBy] as number) - (rowB[sortBy] as number));
+  }
+  // this is a custom type, and can safely assume the gte and lt fields are all numbers or undefined
+  if (type === 'range') {
+    return getRangeCriteria(sortBy, directionFactor);
+  }
+  // IP have a special sorting
+  if (type === 'ip') {
+    return getIPCriteria(sortBy, directionFactor);
+  }
+  // use a string sorter for the rest
+  return (rowA: Record<string, unknown>, rowB: Record<string, unknown>) => {
+    const aString = formatter?.convert(rowA[sortBy]);
```

Comment:
Is it possible there is no formatter? The types indicate otherwise, so if the will always be a formatter, we can remove the optional chaining - if it can actually be undefined, we should do something like this: `aString = formatter ? formatter.convert(rowA[sortBy]) : rowA[sortBy]`, otherwise it will just compare undefined with undefined

Reply:
The `formatFactory` should ensure each column has a formatter:

https://github.com/elastic/kibana/blob/c545b32755dc9bb5fb938af253f49d8055c458a3/x-pack/plugins/lens/public/datatable_visualization/expression.tsx#L128-L130

I'll remove the optional chaining here.